File: ios/pqp/Sources/Core/DeepLink.swift
import Foundation

/// Somewhere in the app a link or a notification is asking to go.
///
/// The same four cases serve three inputs — a universal link, a `pqp://` URL,
/// and the `path` a push payload carries — because all three are the *web
/// client's own routes*. `server/src/services/push.ts` builds that path with
/// `/app/dm/<id>` and `/app/server/<sid>/channel/<cid>`, the web SPA parses it
/// in `client/src/lib/app-route.ts`, and this parses the identical strings. One
/// routing vocabulary, three clients: a change on the server reaches the phone
/// and the browser or neither.
enum DeepLinkTarget: Equatable, Sendable {
    /// An invite code, not yet redeemed. What happens next depends on whether
    /// anybody is signed in — see `SessionStore.handle(_:)`.
    case invite(code: String)
    /// A DM or group conversation: a channel with no server.
    case conversation(channelId: String)
    case channel(serverId: String, channelId: String)
    /// A whole server, with no channel picked out. Where a redeemed invite
    /// lands, since joining tells us the server and nothing else.
    case server(id: String)
}

/// Turns links and notification paths into `DeepLinkTarget`s, and nothing else.
///
/// Deliberately free of any dependency on the session, the network or SwiftUI:
/// the parsing is where the bugs are (a code that arrives percent-encoded, a
/// `pqp://` URL whose first segment is the *host* rather than a path component)
/// and pure functions are the only way to see them before shipping.
enum DeepLink {
    /// The custom scheme, for the contexts where universal links do not fire:
    /// a link typed into Safari's bar, pasted into an app that does not resolve
    /// them, or opened on a device that has not yet fetched the AASA file.
    static let scheme = "pqp"

    /// Invite codes are 8 base64url characters as `generateInviteCode` makes
    /// them (server/src/services/invites.ts). The cap is generous rather than
    /// exact — the server is the authority on whether a code exists, and a
    /// client that refused a code the server would have accepted is a worse
    /// bug than one round trip. What it does is stop a pathological URL from
    /// becoming a pathological request path.
    static let maxInviteCodeLength = 64

    // MARK: - Paths

    /// Parses a web-client route.
    ///
    /// The leading `/app` is optional so the same function reads a push
    /// payload's `/app/dm/x` and a bare `pqp://dm/x`.
    static func target(path: String) -> DeepLinkTarget? {
        var segments = path
            .split(separator: "/", omittingEmptySubsequences: true)
            .map(String.init)
            .compactMap(decode)
        if segments.first == "app" {
            segments.removeFirst()
        }
        // Matched on shape rather than with array patterns, which Swift does not
        // have: `case ["invite", let code]` is not a thing.
        switch segments.count {
        case 2 where segments[0] == "invite":
            let code = segments[1]
            return isUsableCode(code) ? .invite(code: code) : nil

        case 2 where segments[0] == "dm":
            let channelId = segments[1]
            return channelId.isEmpty ? nil : .conversation(channelId: channelId)

        case 2 where segments[0] == "server":
            let serverId = segments[1]
            return serverId.isEmpty ? nil : .server(id: serverId)

        case 4 where segments[0] == "server" && segments[2] == "channel":
            let serverId = segments[1]
            let channelId = segments[3]
            guard !serverId.isEmpty, !channelId.isEmpty else { return nil }
            return .channel(serverId: serverId, channelId: channelId)

        default:
            // `/app` on its own, and anything unrecognised. Both mean "the hub",
            // which is where the app already is — so nothing to navigate to,
            // rather than an error to show.
            return nil
        }
    }

    // MARK: - URLs

    /// Parses either an `https://…/app/invite/CODE` universal link or a
    /// `pqp://invite/CODE` custom-scheme one.
    ///
    /// THE HOST IS PART OF THE PATH UNDER A CUSTOM SCHEME. `pqp://invite/AB12`
    /// has host `invite` and path `/AB12`, so reading `url.path` alone finds
    /// only the code and reading `pathComponents` finds only `["/", "AB12"]`.
    /// Getting this wrong is the classic custom-scheme bug and it presents as
    /// "the link opens the app and then nothing happens".
    ///
    /// The https host is deliberately NOT checked. A universal link can only
    /// arrive from a domain named in the app's Associated Domains entitlement,
    /// so a second check there would be theatre; and for a *pasted* link,
    /// refusing an unfamiliar host would break a staging deploy while
    /// protecting nothing — the code is meaningless except to this app's own
    /// API, which is the thing that decides whether it is real.
    static func target(url: URL) -> DeepLinkTarget? {
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            return nil
        }
        let isCustomScheme = components.scheme?.lowercased() == scheme
        let path = isCustomScheme
            ? "/\(components.host ?? "")\(components.path)"
            : components.path
        return target(path: path)
    }

    // MARK: - Pasted codes

    /// A pasted invite, reduced to the code the API wants — the Swift twin of
    /// `normalizeInviteCode` in `client/src/lib/onboarding.ts`, kept behaviourally
    /// identical because people paste the same links into both clients.
    ///
    /// People paste the whole link, because the whole link is what they were
    /// sent. The last path segment is the code in `/app/invite/<code>`, in
    /// `pqp://invite/<code>`, and in a bare code (which has no segments to drop).
    static func normalizeInviteCode(_ input: String) -> String {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        let withoutQuery = trimmed.split(separator: "?", maxSplits: 1).first
            .map(String.init) ?? ""
        let withoutFragment = withoutQuery.split(separator: "#", maxSplits: 1).first
            .map(String.init) ?? ""
        let last = withoutFragment
            .split(whereSeparator: { $0 == "/" || $0 == "\\" })
            .last
            .map(String.init) ?? ""
        return decode(last) ?? last
    }

    // MARK: - Helpers

    private static func decode(_ segment: String) -> String? {
        segment.removingPercentEncoding ?? segment
    }

    private static func isUsableCode(_ code: String) -> Bool {
        !code.isEmpty && code.count <= maxInviteCodeLength
    }
}

File: android/app/src/main/kotlin/gg/pqp/app/push/DeepLink.kt
package gg.pqp.app.push

/**
 * Where a notification, or a `pqp://` link, wants the app to go.
 *
 * The vocabulary is the web client's routes, not a mobile one, because the
 * server builds exactly one `path` per push and hands the same string to the
 * browser's service worker, to iOS and to here (`buildPushPayload` in
 * `server/src/services/push.ts`). Three clients, one routing vocabulary. This
 * file is the Kotlin half of what `ios/pqp/Sources/Core/DeepLink.swift` does.
 */
sealed interface DeepLinkTarget {
    /** A text channel inside a server. */
    data class Channel(val serverId: String, val channelId: String) : DeepLinkTarget

    /**
     * A DM or a group. `channelId` really is a channel id: a conversation is a
     * `channels` row with `kind` of `dm` or `group`, and `/app/dm/<id>` carries
     * that id, not a separate conversation id.
     */
    data class Conversation(val channelId: String) : DeepLinkTarget

    /** A server with no channel named. */
    data class Server(val serverId: String) : DeepLinkTarget

    /** `pqp://invite/<code>`, which no push produces but the manifest accepts. */
    data class Invite(val code: String) : DeepLinkTarget
}

/**
 * The channel this target is about, or null when it is not about one.
 *
 * Used by [PushPresentation] to answer "is the user already looking at this?",
 * which is the only question that needs a channel id out of a target.
 */
val DeepLinkTarget.channelId: String?
    get() = when (this) {
        is DeepLinkTarget.Channel -> channelId
        is DeepLinkTarget.Conversation -> channelId
        is DeepLinkTarget.Server, is DeepLinkTarget.Invite -> null
    }

object DeepLink {

    /**
     * Parse a server-supplied path into a target, or null if it is not one.
     *
     * THE POINT OF PARSING RATHER THAN LOOKING UP. Everything needed to route
     * is in the string: `/app/server/<sid>/channel/<cid>` carries the server as
     * well as the channel. Resolving the server by searching a channel list the
     * app happens to hold is the bug the web client shipped and fixed in #79,
     * that list only ever contains the *selected* server, so a frame from
     * anywhere else resolved to null, and the notification lost its route, its
     * title and its mute. Nothing here consults app state, which is also why it
     * works from `FirebaseMessagingService` with no session in the process.
     *
     * Accepts an absolute URL (`https://pqp.gg/app/...`), a root-relative path,
     * or a `pqp://` link, because the same parser serves notification taps and
     * the manifest's VIEW intent.
     */
    fun target(path: String?): DeepLinkTarget? {
        val raw = path?.trim().orEmpty()
        if (raw.isEmpty()) return null

        // `pqp://invite/CODE` has its code in the *authority*, not the path, so
        // it is matched before anything splits on slashes.
        pqpScheme(raw)?.let { return it }

        val segments = pathSegments(raw)
        if (segments.isEmpty()) return null

        if (segments[0] == "invite" && segments.size >= 2) {
            return if (isUsableCode(segments[1])) DeepLinkTarget.Invite(segments[1]) else null
        }
        if (segments[0] != "app") return null

        return when {
            segments.size >= 5 &&
                segments[1] == "server" &&
                segments[3] == "channel" ->
                DeepLinkTarget.Channel(segments[2], segments[4])

            segments.size >= 3 && segments[1] == "server" ->
                DeepLinkTarget.Server(segments[2])

            segments.size >= 3 && segments[1] == "dm" ->
                DeepLinkTarget.Conversation(segments[2])

            else -> null
        }
    }

    private fun pqpScheme(raw: String): DeepLinkTarget? {
        if (!raw.startsWith(PQP_SCHEME)) return null
        val rest = raw.removePrefix(PQP_SCHEME).trim('/')
        if (rest.isEmpty()) return null
        val parts = rest.split('/').filter { it.isNotEmpty() }
        return when {
            parts.size >= 2 && parts[0] == "invite" && isUsableCode(parts[1]) ->
                DeepLinkTarget.Invite(parts[1])
            else -> null
        }
    }

    /**
     * What an invite code may contain, before it is put in a URL path.
     *
     * Any app on the phone can fire a `pqp://` intent at this activity, so the
     * segment that becomes `POST /api/invites/<code>/join` is checked here
     * rather than trusted. The generator makes short alphanumeric codes
     * (`server/src/services/invites.ts`); the cap is generous rather than
     * exact, which is the same call `ios/pqp/Sources/Core/DeepLink.swift`
     * makes. The Swift twin of this function.
     */
    private fun isUsableCode(code: String): Boolean =
        code.length in 1..64 && code.all { it.isLetterOrDigit() || it == '-' || it == '_' }

    /**
     * The path, minus scheme, host, query and fragment, split on `/`.
     *
     * Hand-rolled rather than `Uri.parse`, because this runs in unit tests on a
     * plain JVM where `android.net.Uri` is an unimplemented stub that returns
     * null for everything. The grammar being parsed is four fixed shapes, not
     * RFC 3986.
     */
    private fun pathSegments(raw: String): List<String> {
        var value = raw
        val schemeEnd = value.indexOf("://")
        if (schemeEnd >= 0) {
            val afterScheme = value.substring(schemeEnd + 3)
            val slash = afterScheme.indexOf('/')
            value = if (slash < 0) "" else afterScheme.substring(slash)
        }
        value = value.substringBefore('?').substringBefore('#')
        return value.split('/').filter { it.isNotEmpty() }
    }

    private const val PQP_SCHEME = "pqp://"
}
